src/domain/datasets/services/src/use_cases/versioned_files/update_version_file_use_case_impl.rs

```diff
@@ -0,0 +1,227 @@
+    #[tracing::instrument(level = "info", name = UpdateVersionFileUseCaseImpl_execute, skip_all, fields(%dataset_handle.id))]
+    async fn execute(
+        &self,
+        dataset_handle: &odf::DatasetHandle,
+        content_args_maybe: Option<ContentArgs>,
+        expected_head: Option<odf::Multihash>,
+        extra_data: Option<ExtraDataFields>,
+    ) -> Result<UpdateVersionFileResult, UpdateVersionFileUseCaseError> {
+        let resolved_dataset = self
+            .rebac_dataset_registry_facade
+            .resolve_dataset_by_ref(&dataset_handle.as_local_ref(), auth::DatasetAction::Write)
+            .await
+            .map_err(|e| {
+                use RebacDatasetRefUnresolvedError as E;
+                match e {
+                    E::NotFound(e) => UpdateVersionFileUseCaseError::NotFound(e),
+                    E::Access(e) => UpdateVersionFileUseCaseError::Access(e),
+                    e @ E::Internal(_) => UpdateVersionFileUseCaseError::Internal(e.int_err()),
+                }
+            })?;
+
+        let entity = if let Some(args) = content_args_maybe {
+            let (latest_version, _) = self
+                .get_latest_version(&resolved_dataset.get_handle().as_local_ref())
+                .await?;
+            let new_version = latest_version + 1;
+
+            let result = VersionedFileEntity::new(
+                new_version,
+                args.content_hash.clone(),
+                args.content_length,
+                args.content_type,
+                extra_data,
+            );
+
+            // Upload data object in case when content is present
+            if let Some(content_stream) = args.content_stream {
+                let data_repo = resolved_dataset.as_data_repo();
+                data_repo
+                    .insert_stream(
+                        content_stream,
+                        odf::storage::InsertOpts {
+                            precomputed_hash: Some(&result.content_hash),
+                            size_hint: Some(result.content_length as u64),
+                            ..Default::default()
+                        },
+                    )
+                    .await
+                    .int_err()?;
+            }
+
+            result
+        } else {
+            let mut last_entity = self
+                .get_versioned_file_entity_from_latest_entry(
+                    &dataset_handle.as_local_ref(),
+                    extra_data,
+                )
+                .await?
+                .unwrap();
+
+            // Increment version to match next record value
+            last_entity.version += 1;
+
+            last_entity
+        };
+
+        if entity.content_length > self.upload_config.max_file_size_in_bytes() {
+            return Err(UpdateVersionFileUseCaseError::TooLarge(
+                UploadTooLargeError::new(
+                    entity.content_length,
+                    self.upload_config.max_file_size_in_bytes(),
```

Comment:
I only see one size limit but I think we need two:
- Overall file content limit
- Limit for in-band upload (see my other comments for ideas on this)